storm-server/src/main/java/org/apache/storm/metricstore/rocksdb/RocksDbStore.java

```diff
@@ -0,0 +1,636 @@
+        metricsWriter = new RocksDbMetricsWriter(this, this.queue, this.failureMeter);
+
+        int cacheCapacity = Integer.parseInt(config.get(DaemonConfig.STORM_ROCKSDB_METADATA_STRING_CACHE_CAPACITY).toString());
+        StringMetadataCache.init(metricsWriter, cacheCapacity);
+        readOnlyStringMetadataCache = StringMetadataCache.getReadOnlyStringMetadataCache();
+        metricsWriter.init(); // init the writer once the cache is setup
+
+        // start threads after metadata cache created
+        Thread thread = new Thread(metricsCleaner, "RocksDbMetricsCleaner");
+        thread.start();
+        thread = new Thread(metricsWriter, "RocksDbMetricsWriter");
+        thread.start();
+    }
+
+    /**
+     * Implements configuration validation of Metrics Store, validates storm configuration for Metrics Store.
+     *
+     * @param config Storm config to specify which store type, location of store and creation policy
+     * @throws MetricException if there is a missing required configuration or if the store does not exist but
+     *                         the config specifies not to create the store
+     */
+    private void validateConfig(Map config) throws MetricException {
+        if (!(config.containsKey(DaemonConfig.STORM_ROCKSDB_LOCATION))) {
+            throw new MetricException("Not a vaild RocksDB configuration - Missing store location " + DaemonConfig.STORM_ROCKSDB_LOCATION);
+        }
+
+        if (!(config.containsKey(DaemonConfig.STORM_ROCKSDB_CREATE_IF_MISSING))) {
+            throw new MetricException("Not a vaild RocksDB configuration - Does not specify creation policy "
+                    + DaemonConfig.STORM_ROCKSDB_CREATE_IF_MISSING);
+        }
+
+        // validate path defined
+        String storePath = getRocksDbAbsoluteDir(config);
+
+        boolean createIfMissing = ObjectReader.getBoolean(config.get(DaemonConfig.STORM_ROCKSDB_CREATE_IF_MISSING), false);
+        if (!createIfMissing) {
+            if (!(new File(storePath).exists())) {
+                throw new MetricException("Configuration specifies not to create a store but no store currently exists at " + storePath);
+            }
+        }
+
+        if (!(config.containsKey(DaemonConfig.STORM_ROCKSDB_METADATA_STRING_CACHE_CAPACITY))) {
+            throw new MetricException("Not a valid RocksDB configuration - Missing metadata string cache size "
+                    + DaemonConfig.STORM_ROCKSDB_METADATA_STRING_CACHE_CAPACITY);
+        }
+
+        if (!config.containsKey(DaemonConfig.STORM_ROCKSDB_METRIC_RETENTION_HOURS)) {
+            throw new MetricException("Not a valid RocksDB configuration - Missing metric retention "
+                    + DaemonConfig.STORM_ROCKSDB_METRIC_RETENTION_HOURS);
+        }
+    }
+
+    private String getRocksDbAbsoluteDir(Map conf) throws MetricException {
+        String storePath = conf.get(DaemonConfig.STORM_ROCKSDB_LOCATION).toString();
+        if (storePath == null) {
```

Comment:
storePath should never be null because we called toString on the original object just one line above.  We probably want to cast it to a String instead of calling toString on it.